clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,209 @@
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// createKSMLatencyReport gathers the latency one last time and creates the summary based on the Quantile from the sub histograms.
+// Afterwards it creates the Summary Report.
+func (m *ksmLatencyMeasurement) createKSMLatencySummary(ctx context.Context, client clientset.Interface) ([]measurement.Summary, error) {
+	if !m.isRunning {
+		klog.V(2).Infof("%s: measurement was not running", m)
+		return nil, nil
+	}
+	close(m.stopCh)
+	m.wg.Wait()
+	latestLatency, err := m.retrieveKSMLatencyMetrics(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	// We want to subtract the latest histogram from the first one we collect.
+	finalLatency := HistogramSub(latestLatency, m.initialLatency)
+	// Pretty Print the report.
+	result := &measurementutil.LatencyMetric{}
+	err = SetQuantileFromHistogram(result, finalLatency)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	// Create Summary.
+	return []measurement.Summary{measurement.CreateSummary(ksmLatencyName, "json", content)}, nil
+}
+
+// startQuerying queries /metrics endpoint of kube-state-metrics kube_ metrics every interval
+// and stops when stop is called.
+func (m *ksmLatencyMeasurement) startQuerying(ctx context.Context, client clientset.Interface, interval time.Duration) {
+	m.isRunning = true
+	m.wg.Add(1)
+	go m.queryLoop(ctx, client, interval)
+}
+
+func (m *ksmLatencyMeasurement) queryLoop(ctx context.Context, client clientset.Interface, interval time.Duration) {
+	defer m.wg.Done()
+	for {
+		select {
+		case <-m.stopCh:
+			return
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+			var output string
+			output, err := m.getMetricsFromService(ctx, client, m.metricsPort)
+			if err != nil {
+				klog.V(2).Infof("%s: measurement already running", m)
+			}
+			if output == "" {
+				klog.V(2).Infof("%s: measurement already running", m)
+				klog.V(2).Infof("/metrics endpoint of kube-state-metrics returned no data in namespace: %s from service: %s port: %d", m.namespace, m.serviceName, m.metricsPort)
+			}
+
+		}
+	}
+}
+
+func (m *ksmLatencyMeasurement) retrieveKSMLatencyMetrics(ctx context.Context, c clientset.Interface) (*measurementutil.Histogram, error) {
+	ksmHist := measurementutil.NewHistogram(nil)
+	output, err := m.getMetricsFromService(ctx, c, m.selfPort)
+	if err != nil {
+		return ksmHist, err
+	}
+	samples, err := measurementutil.ExtractMetricSamples(output)
+	if err != nil {
+		return ksmHist, err
+	}
+	for _, sample := range samples {
+		switch sample.Metric[model.MetricNameLabel] {
+		case ksmRequestDurationMetricName:
+			measurementutil.ConvertSampleToHistogram(sample, ksmHist)
+		}
+	}
+	return ksmHist, nil
+}
+
+func (m *ksmLatencyMeasurement) getMetricsFromService(ctx context.Context, client clientset.Interface, port int) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```

Comment:
Isn't 2 minutes too high? Can we align this with the 30s interval period?

Reply:
Yes 2 minutes is above the 30 second scrape interval we recommend, but timeout should be a bit above 30 seconds no? So 40 seconds? Not sure if that makes sense.

cc @brancz any suggestions?

Reply:
If we want to align it with the interval, then I'd actually prefer to make the interval 2m as well. The point of this is kind of to see how it behaves and if it's over 30s then we'd want to understand how much over 30s it is (this is effectively a bulk endpoint so 30s is not unthinkable in extremely huge clusters). 2m is so high that if it reaches that value it should legitimately just fail.

Reply:
OK - we can tune that later too.

Reply:
Sounds good to me!